docs/source/daml/reference/generic-templates.rst

```diff
@@ -0,0 +1,60 @@
+.. Copyright (c) 2019 The DAML Authors. All rights reserved.
+.. SPDX-License-Identifier: Apache-2.0
+
+
+Generic Templates
+#################
+
+Sometimes different DAML templates have a common structure. Typically this occurs when there is some logic (usually in choices) that can be applied to many different underlying contracts, such as assets. Generic templates allow you to abstract template code from some of the data that would occur in the contracts. Let's see a simple example to understand what this means.
+
+Example: Generic Proposal
+*************************
+
+Suppose we want to model a propose and accept workflow. This means that a party can propose a contract to some number of receiving parties, and any one of them can accept it with the terms of that contract. We see this pattern occur frequently for different types of contracts. Of course, we could implement one proposal template for every underlying template. However this is tedious and error prone. Alternatively, we can write it once and for all using a generic template.
+
+This is how a generic Proposal template looks in DAML.
+
+.. literalinclude:: ../code-snippets/Proposal.daml
+   :language: daml
+   :start-after: -- start proposal body
+   :end-before: -- end proposal body
+
+There are several things to notice in this short template.
+
+Firstly, where there is usually just a template name, there is now a *template header* ``Template t => Proposal t``. Here ``Proposal`` takes a single type parameter ``t`` representing the type of the underlying asset. The ``Template`` constraint says that ``t`` is not just any type but a contract template with signatories, choices, etc.
```

Comment:
`template header` sounds a bit weird to me but I don’t have a better proposal (no pun intended) sadly. I think it might also be worth pointing out that this is not something completely different but simply a generalization of the simple case where there is only a template name.